rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,288 @@
+
+Let's take the end of life of Python 2 as an example. (This applies to other ecosystems as well, and will come up again and again in the future.) It has sparked a few bulk package removal actions by dedicated persons, but those are pretty work intensive and prone to burn out maintainers. A goal of this RFC is to provide a way to notify all users of a package about the outstanding issues. This will hopefully draw more attention to abandoned packages, and spread the work load. It can also help soften the removal of packages by providing a period for users to migrate away at their own pace.
+
+For some use cases, like for packages without maintainers, we do not want to break evaluation of a package and simply warn the user instead. We want the users to configure all these according to their needs and through a single standardized interface.
+
+## Detailed design
+[design]: #detailed-design
+
+### Package problems
+
+A new attribute is added to the `meta` section of a package: `problems`. If present, it is a an attribute set of attrsets which each have at least the following fields:
+
+- `kind`: Required. The resulting warning will be printed as `kind: message`. Defaults to the attrset key if absent.
+- `message`: Required. A string message describing the issue with the package. The value should:
+  - Start with the "This package", "The application" or equivalent, or simply with the package name.
+  - Be capitalized (unless it starts with the package name).
+  - Use a period at the end.
+- `name`: Inferred from the attrset key if `kind` is specified. Give the issue a custom name for more easy filtering.
+- `date`: Required. An ISO 8601 `yyyy-mm-dd`-formatted date from when the issue was added.
+- `urls`: Optional, list of strings. Can be used to link issues, pull requests and other related items.
+
+Other attributes are allowed. Some message kinds may specify additional required attributes.
+The keys in the attribute set are used to infer the `name` or `kind` values within, depending on which is present.
+
+Example values:
+
+```nix
+meta.problems = {
+  # This one will have the name "python2-eol"
+  python2-eol = {
+    kind = "deprecated";
+    message = "This package depends on Python 2, which has reached end of life.";
+    date = "1970-01-01";
+    urls = [ "https://github.com/NixOS/nixpkgs/issues/148779" ];
+  };
+  # This one will have no name
+  removal = {
+    # kind = "removal"; # Inferred from attribute key
+    message = "The application has been abandoned upstream, use libfoo instead";
+    date = "1970-01-01";
+  };
+};
+```
+
+### Problem kinds
+
+At the moment, the following values for the `kind` field of a warning are known:
+
+- `removal`: The package is scheduled for removal some time in the future.
+- `deprecated`: The package has been abandoned upstream or has end of life dependencies.
```

Comment:
The general concept here is correct, although I agree that the name may be misleading. The representative use case for "deprecated" is the Python 2 EOL: some packages can be fixed by upgrading to Python 3 or otherwise removing the Python 2 dependency, others are probably hopeless and doomed.

> they can still be maintained while having EOL dependencies.

IMO it's the other way around: packages with EOL dependencies are in need of maintenance

Reply:
Also, EOL packages receive security fixes or maintenance only via paid subscriptions usually.

Reply:
Hello, came to read the rfc after rcp was announced, it's the only thing that bugged me while reading..

It seems there was no conclusion to this comment.

The distinction can be hard to grasp, but I think deprecated and eol/obsolete are clearly different and should be handled differently. 

Mozilla's JS docs uses deprecated/obsolete [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Deprecated_and_obsolete_features):
> **deprecated**: still available but planned for removal
> **obsolete**: no longer usable

To me (french, not english-native), obsolete also means that something else took its place. `eol` would be stronger and wouldn't mean that.

Random guy on english stackexchange [says](https://english.stackexchange.com/a/87891):
> **Deprecated** is more or less a "marker", saying that it should not be used, something else that has the same effect has been created, and it is soon to be deleted. It may still work as expected (read the last paragraph on why I say may), but it will vanish soon.
> **Obsolete** means that it no longer works as expected, or doesn't do anything at all.

Reply:
@bew, I don't see how this distinction is relevant in this context. The idea behind "deprecation" is, that some package has a problem related to software for which upstream dropped support, and thus needs fixing or removal. Whether or not it currently works is not relevant to me, since this is from a maintainer perspective and not a user perspective.

Reply:
While I wanted also a distinction, I feel like it's not that important from a maintainer perspective, and I agree with @bew ; though, if, in practice, we found ourselves wanting more, we can always discuss this after extensive experience with that RFC in production.

Reply:
Ok for tje distinction between deprecated/obsolete/eol.

I'm now also thinking about the wording vs semantic difference between `removal` and `deprecated`.

Something that is marked for `removal` should not be used anymore, this actually looks like the definition of `deprecated` 🤔

Might be confusing, maybe we can rename:
- `removal` to `deprecated` (or don't use `deprecated` at all)
- `deprecated` to `obsolete`/`eol`

What do you think?

Reply:
I'm fine with changing names as long as we keep the originally intended semantics. I'd prefer to keep "removal" because it's for things "that are about to be removed" so I find it quite fitting. "removal" is a least resort measure, and "deprecated" would be too weak of a word for it.

I'd be fine with renaming "deprecated" to "obsolete" or "eol". Personally I don't mind "deprecated" and don't think the alternatives are a big improvement, but they're fine so if people prefer it we can do that.